core/txpool/blobpool/blobpool.go

```diff
@@ -0,0 +1,1461 @@
+// kept in sync with the main transacion pool's internal state.
+func (p *BlobPool) Reset(oldHead, newHead *types.Header) {
+	waitStart := time.Now()
+	p.lock.Lock()
+	resetwaitHist.Update(time.Since(waitStart).Nanoseconds())
+	defer p.lock.Unlock()
+
+	defer func(start time.Time) {
+		resettimeHist.Update(time.Since(start).Nanoseconds())
+	}(time.Now())
+
+	statedb, err := p.chain.StateAt(newHead.Root)
+	if err != nil {
+		log.Error("Failed to reset blobpool state", "err", err)
+		return
+	}
+	p.head = newHead
+	p.state = statedb
+
+	// Run the reorg between the old and new head and figure out which accounts
+	// need to be rechecked and which transactions need to be readded
+	if reinject, inclusions := p.reorg(oldHead, newHead); reinject != nil {
+		for addr, txs := range reinject {
+			// Blindly push all the lost transactions back into the pool
+			for _, tx := range txs {
+				p.reinject(addr, tx)
+			}
+			// Recheck the account's pooled transactions to drop included and
+			// invalidated one
+			p.recheck(addr, inclusions)
+		}
+	}
+	// Flush out any blobs from limbo that are older than the latest finality
+	p.limbo.finalize(p.chain.CurrentFinalBlock())
+
+	// Reset the price heap for the new set of basefee/blobfee pairs
+	var (
+		basefee = uint256.MustFromBig(misc.CalcBaseFee(p.chain.Config(), newHead))
+		blobfee = uint256.MustFromBig(big.NewInt(params.BlobTxMinDataGasprice))
+	)
+	if newHead.ExcessDataGas != nil {
+		blobfee = uint256.MustFromBig(misc.CalcBlobFee(*newHead.ExcessDataGas))
+	}
+	p.evict.reinit(basefee, blobfee, false)
+
+	basefeeGauge.Update(int64(basefee.Uint64()))
+	blobfeeGauge.Update(int64(blobfee.Uint64()))
+	p.updateStorageMetrics()
+}
+
+// reorg assembles all the transactors and missing transactions between an old
+// and new head to figure out which account's tx set needs to be rechecked and
+// which transactions need to be requeued.
+//
+// The transactionblock inclusion infos are also returned to allow tracking any
+// just-included blocks by block number in the limbo.
+func (p *BlobPool) reorg(oldHead, newHead *types.Header) (map[common.Address][]*types.Transaction, map[common.Hash]uint64) {
+	// If the pool was not yet initialized, don't do anything
+	if oldHead == nil {
+		return nil, nil
+	}
+	// If the reorg is too deep, avoid doing it (will happen during snap sync)
+	oldNum := oldHead.Number.Uint64()
+	newNum := newHead.Number.Uint64()
+
+	if depth := uint64(math.Abs(float64(oldNum) - float64(newNum))); depth > 64 {
+		return nil, nil
+	}
+	// Reorg seems shallow enough to pull in all transactions into memory
+	var (
+		transactors = make(map[common.Address]struct{})
+		discarded   = make(map[common.Address][]*types.Transaction)
+		included    = make(map[common.Address][]*types.Transaction)
+		inclusions  = make(map[common.Hash]uint64)
+
+		rem = p.chain.GetBlock(oldHead.Hash(), oldHead.Number.Uint64())
+		add = p.chain.GetBlock(newHead.Hash(), newHead.Number.Uint64())
+	)
+	if rem == nil {
```

Comment:
should we also check the nilness of `add`? Theoretically it won't be nil, but just in case...

Reply:
Yeah, I guess we did add the same check to the legacy pool recently: https://github.com/ethereum/go-ethereum/commit/e4660a1181d19c6c50ebf77eb3ff3865948c2d63